storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java

```diff
@@ -183,6 +197,21 @@ public boolean isFinalizedEpoch(final UnsignedLong epoch) {
     return historicalChainData.getFinalizedStateByBlockRoot(blockRoot);
   }
 
+  public BeaconState regenerateBeaconState(final BeaconState preState, final UnsignedLong slot) {
```

Comment:
I suspect rather than this being a public method, we should just expose a method like `getExactStateAtSlot(slot)` which gets the block from the previous block and automatically processes empty slots as needed.  Saves having to reimplement that logic in multiple places.

Reply:
yep fair call, it'll clean up the provider a bit.